Feature/hmmsearch

aggreen opened the pull request:
This adds the feature to search the PDB database by HMM build and search, rather than by Jackhmmer. 

Branch cloned from ChannyClaus/cleanup2 and Thomas's requested changes from October 30th were made. The only missing feature is how to properly number the hits resulting from the HMM to ensure accurate residue mapping between query and sequence, see comments below.

aggreen commented:
Comments on original pull request (ChannyClaus/cleanup2) from Thomas on October 30th re HMM numbering: 

---
Also, there is one conceptual complication that I didn't anticipate, which needs to be addressed before this can work properly:

an HMM-based search protocol which starts from an input alignment should not need a query sequence like the jackhmmer-based protocol (this code is still in the hmmsearch protocol but doesn't make much sense)
The alignment coming out of hmmsearch will not contain the aligned query sequence like the jackhmmer-based alignment (or is there any information that goes beyond the HMM numbering which presumably starts at 1? Please check this on an example). This information is crucial for establishing the residue-level mapping from query to aligned sequences in structure index mapping
So somehow the query sequence + region has to be in the final alignment coming out of the protocol.

The information about the query sequence is contained in the alignment that we put into hmmbuild in the first place, so I think there is a way round this issue:

make sure match states of the HMM correspond 1:1 to residues of query sequence
put query sequence into alignment (first sequence), filling any non-match column with gaps.
Not particularly elegant, but I can't think of a better way of doing this right now. Any other ideas?

---

My reply on November 10th:

---

I think your solution for ensuring correct numbering is a fine one. If we generate the hmm from the raw_focus_alignment_file, there is a one-to-one correspondence between positions in the hmm and positions in the query sequence. Then, it will be a matter of parsing the output .sto file for each position that has an x in the RF field (fields that matched the hmm) and only including those positions in the parsed fasta alignment. I could write a dumb function to do this, but is there any existing functionality? See attached .sto, .fa, and raw.fasta files.

I believe that this also creates an argument for keeping the functionality that creates the target sequence file at the beginning of hmm_build_and_search, since we'll want a source of the target sequence for alignment later. I've left that code block in for now for this reason.

[DIVIB_BACSU_b0.2_raw.fasta.txt](https://github.com/debbiemarkslab/EVcouplings/files/1467533/DIVIB_BACSU_b0.2_raw.fasta.txt)
[DIVIB_BACSU_b0.2.fa.txt](https://github.com/debbiemarkslab/EVcouplings/files/1467534/DIVIB_BACSU_b0.2.fa.txt)
[DIVIB_BACSU_b0.2.sto.txt](https://github.com/debbiemarkslab/EVcouplings/files/1467535/DIVIB_BACSU_b0.2.sto.txt)

---

aggreen commented:
I have looked more into solving this and it is actually more complicated than I had initially thought. 

Turns out, HMMbuild internally decides which columns in the MSA are 'consensus' columns that should be included HMM, and these columns are not guaranteed to be continuous. If we want to continue to use that default behavior, we'll have to write a function to parse the .hmm file to extract its mapping from hmm state to alignment column, then create a function to write the new alignment, adding gaps as necessary. Then, we have to fix hit table variable create right now in find_homologs. Since the ali_to and ali_from coordinates are not meaningful for HMMsearch, we'd have to get the hmm hit coordinates instead. 

This solution isn't too bad, except I fear that parsing the .hmm file is going to be messy and that we might introduce the possibility for a lot of weird edge cases...

An easier but potentially worse solution would be to use the non-default HMMbuild options and coerce every column to be a consensus column. Then we don't have to bother with parsing the .hmm file as there will be a one-to-one correspondence between the HMM match states and the query sequence.

aggreen commented:
Alright, I have solved this problem as follows:

1. Opted to coerce HMMbuild to use all of the columns in the input alignment as match states. I did this after Perry pointed out that if you don't coerce this option, if the alignment has gone too far in sequence space you will end up with an HMM that has very little to do with your query sequence at all. I had to create a new global variable called SYMFRAC_HMMBUILD in the align.tools file which must be set to 0.0 - this allows all columns into the HMM. If this variable is changed the mapping code will break, which I don't like, but that's why I didn't create a new parameter for the user to manipulate. 

2. Wrote code that creates a raw.fasta file that contains the query sequence with added gaps for non-match states, and all of the hit sequences. We had to add the gaps to the query sequence in order to preserve the continuity of the hit sequences for mapping purposes.

thomashopf commented:
Here's a few points from a very quick glance-over:

* _make_hmmsearch_raw_fasta would be better in the align protocol (i.e. an alignment that comes out of that protocol always has the query as first sequence)
* I would move the SYMFRAC_HMMBUILD constant inside the respective protocol (in a sense that is an implementation detail to that particular protocol)
* hmmbuild_and_search should not require an input target sequence but rather extract that sequence from the alignment (sorry to be such a pest about this :)) based on identifier or position (like first sequence)  - just like the existing() protocol does from an existing alignment, as all the information is in the input sequence alignment, and the behaviour should be consistent

Re the match state problem: When using --symfrac 0, won't that include all sorts of other insert positions relative to the query unless you put in the focus alignment? 

I just saw that hmmuild has an option --hand. Wouldn't selecting all the columns corresponding to query positions be the safest solution?

Also should be viewed in context of #87 since this would plug in at the very same point in compare (if we manage to get a specifically constructed database with Uniprot sequences corresponding to PDB sequences)

aggreen commented:
Okay I have finally gotten around to addressing these:

1. I moved _make_hmmsearch_raw_fasta to align/protocol, though I'm not sure why it's better there as I only see it being used after calling run_hmmbuild

2.  I created a new function (protocol/_format_alignment_for_hmmbuild) that is called before running hmmbuild to ensure the alignment is properly formatted and get the target sequence. The function takes in a sequence alignment, makes sure the target sequence is present, make sure the alignment is in focus mode (see point # 4) and moves the target sequence to the first index. If the user is inputting a raw_focus_alignment_file from the pipeline, this shouldn't change the input alignment at all, but it is a good guarantee.

3. I moved the SYMFRAC parameter to inside the function

4. Regarding using -hand versus -symfrac, -hand requires an input .sto formatted sequence alignment with a # RF annotation line where the reference columns exactly match the query sequence. I'm not sure if this is guaranteed to be the case with all output .sto files from jackhmmer. If it is guaranteed, I'm still not convinced that having hmmbuild_and_search dependent on having a .sto input is a good idea, since we like to delete the .sto files. I like the symfrac option better so that re-running the compare stage is possible without a .sto file, and with the new _format_alignment_for_hmmbuild function this will be guaranteed if the user runs run_hmmbuild_and_search. I put a warning in the run_hmmbuild documentation.

5. I fixed the code to actually handle a case where we find no structures :)  This is done in _make_hmmsearch_raw_fasta where if the .sto alignment provided is empty, we just return the target sequence file.

aggreen commented:
FYI, my new pull request has a lot of commits that look repeated from before. I think this is because I rebased from debbiemarkslab/master before making the new changes so I think it counts all the commits that I had to replay.

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+    def _add_gaps_to_query(query_sequence_ali, ali):
+
+        new_sequence = ""
+        seq = list(query_sequence_ali .matrix[0, :])
```
no space before .matrix

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+        seq = list(query_sequence_ali .matrix[0, :])
+
+        # loop through every position in the HMMsearch hits
+        for i in range(0, len(ali.annotation["GC"]["RF"])):
```
1) would be safer to check here if annotation actually exists and to raise a meaningful exception if not

2) range(0, ..) should just be range(...)

aggreen replied:
We already checked if the annotation exists in lines 105-109, and raise a ValueError if not

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+    """
+    def _add_gaps_to_query(query_sequence_ali, ali):
+
+        new_sequence = ""
```
would call this gapped_query_sequence

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+
+    # open the sequence file
+    with open(ar["target_sequence_file"]) as a:
+        query_sequence_ali = Alignment.from_file(a, format='fasta')
```
" instead of '

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+
+    # if the provided alignment is empty, just return the target sequence 
+    if not valid_file(ar["raw_alignment_file"]):
+
```
remove empty line

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+
+    # else, open the HMM search result
+    with open(ar["raw_alignment_file"]) as a:
+        ali = Alignment.from_file(a, format='stockholm')
```
" instead of '

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+        )
+            
+    # get the index of columns that do not contain match states (indicated by an x)
+    gap_index = [i for i, x in enumerate(ali.annotation["GC"]["RF"]) if x != "x"]
```
line wrap would be nice here and in line below
```
gap_index = [
    i for i, x in enumerate(ali.annotation["GC"]["RF"]) if x != "x"
]
```

thomashopf commented on evcouplings/align/protocol.py:
```diff
@@ -38,6 +40,106 @@
+
+    # add insertions to the query sequence in order to preserve correct
+    # numbering of match sequences
+    gapped_sequence_ali = _add_gaps_to_query(query_sequence_ali, ali)
```
the code determining match indices seems a bit redundant to what happens inside _add_gaps_to_query, maybe move that check inside the sub-function?

thomashopf commented:
Thanks Anna, this looks very nice!

Regarding the general points above:
1) The idea behind moving this function (not phrased very clearly I see now) was that hmmbuild_and_search would return an alignment that already contains the query sequence, rather than patching the query in the compare stage (as is the current implementation). Goal would be to make the hmmbuild_and_search protocol a completely contained box where you go in with a sequence alignment containing the query -> HMM magic -> sequence alignment against other database containing the query comes out

2) Yep elegant solution!

4) I agree, definitely would be a bad idea to require a .sto alignment here. I was thinking more along the lines of creating a temporary .sto alignment inside the function that contains the annotation and feeding that into hmmbuild. This would be the 100% clean solution to get proper insertion and deletion transition probabilities but I am really not sure if worth the effort - if someone wrote a .sto writer function (which should be relatively straightforward), it would be a quick thing to implement.

5) Good catch...

Two other things that came to mind:
1) Effectively replacing the Orchestra v1 config file with the v2 config file is on purpose I assume? (Which I think is good given that v1 is going to be retired)

2) Are there any test runs of the HMM-based code with contacts maps to look at? I think we should test and compare on a handful examples, just to be on the safe side; including some pathological PDB cases

aggreen commented on evcouplings/align/protocol.py:
```diff
@@ -1013,6 +1115,257 @@ def jackhmmer_search(**kwargs):
+
+    # prepare input alignment for hmmbuild
+    focus_fasta_file, target_sequence_file, region_start, region_end = \
+    _format_alignment_for_hmmbuild(
```
you're right - looks like pep8 wants you to indent after using a backslash to continue a line

aggreen commented on evcouplings/align/protocol.py:
```diff
@@ -1013,6 +1115,257 @@ def jackhmmer_search(**kwargs):
+
+        # save results of search for possible restart
+        write_config_file(ali_outcfg_file, ali)
+
```
note to self: fixed all comments up to (not including) this on 1/3

aggreen replied:
Actual reply: it would be easy to patch in the query sequence if I can go from .sto input -> fasta output. Otherwise I will have to get around to writing the .sto writer...

aggreen commented:
Original points:

1. As I said above, this would be easy if you'll forgive me for outputting a .sto that has no query sequence along with a .fasta alignment that has the query sequence. I'm trying to avoid writing the .sto writer as I feel I've already sunk too much time into this feature. 

Other points:

1. This deletion was actually due to rebasing hell. I will put the v1 file back. 

2. I have run ~40 proteins with this hmmsearch method and in general it seems to work well. There are several cases where jackhmmer doesn't find a homolog that hmmsearch does find. The only pathological case I have found so far is at bitscore 0.1, where we have a giant sequence alignment that has diverged from the query sequence, then the hmm will find structures that correspond to this divergent alignment rather than to the actual query. In those cases you find high precision even though you haven't actually modeled the query sequence. would you like me to share these examples and if so, how?

thomashopf commented:
I think it is okay without the Stockholm writer for now... let's just create a ticket to eventually add it (at that point, we could also think about the --hand option).

Don't want to create any more work here :)

aggreen commented:
Okay, got it to work without the stockholm writer, restored the old config files, and created a new issue

b-schubert commented:
Please PR on develop now
